EIPS/eip-2124.md

```diff
@@ -0,0 +1,294 @@
+| A |  | A ⊕ B |  | No (4) | Local needs software update. |
+| A | B | A ⊕ B ⊕ C | | No² (4) | Local needs software update. |
+| [A,B] |  | A |  | No (4) | Remote needs software update. |
+
+*Note, there's one asymmetry in the table, marked with ¹ and ². Since we don't have access to a remote node's future fork list (just the next one), we can't detect that it's software is stale until it syncs up. This is acceptable as 1) the remote node will disconnect from us anyway, and 2) this is a temporary fluke during sync, not permanent with a leftover node.*
+
+#### Mismatching chain examples (local perspective)
+
+TODO: Give some examples as to what happens if the nodes follow different forks.
+
+## Rationale
+
+##### Why flatten `FORK_HASH` into 4 bytes? Why not share the entire genesis and fork list?
+
+Whilst the `eth` devp2p protocol permits arbitrarily much data to be transmitted, the discovery protocol relies on MTU-limited UDP messages. The total space allowance for all ENR entries is 300 bytes.
+
+Reducing the `FORK_HASH` into a 4 bytes checksum ensures that we leave ample room in the ENR for future extensions; and 4 bytes is more than enough for arbitrarily many Ethereum networks from a (practical) collision perspective.
+
+##### Why use IEEE CRC32 as the checksum instead of Keccak256?
+
+We need a mechanism that can flatten arbitrary data into 4 bytes, without ignoring any of the input. Any other checksum or hashing algorithm would work, but since nodes can lie at any time, there's no value in cryptographic hash functions.
+
+Instead of just taking the first 4 bytes of a Keccak256 hash (seems odd) or XOR-ing all the 4-byte groups (messy), CRC32 is a better alternative, as this is exactly what it was designed for. IEEE CRC32 is also used by ethernet, gzip, zip, png, etc, so every programming language support should not be a problem.
+
+##### We're not using `FORK_NEXT` for much, can't we get rid of it somehow?
+
+We need to be able to differentiate whether a remote node is out of sync or whether its software is stale. Sharing only the past forks cannot tell us if the node is legitimately behind or stuck.
+
+##### Why advertise only one next fork, instead of "hashing" all known future ones like the `FORK_HASH`?
+
+Opposed to past forks that have already passed (for us locally) and can be considered immutable, we don't know anything about future ones. Maybe we're out of sync or maybe the fork didn't pass yet. If it didn't pass yet, it might be postponed, so enforcing it would split the network apart. It could also happen that we're not yet aware of all future forks (haven't updated our software in a while).
+
+## Backwards Compatibility
+
+This EIP only defines an identity scheme, it does not define functional changes.
+
+## Test Cases
+
+Here's a full suite of tests for all possible fork IDs that Mainnet, Ropsten, Rinkeby and Görli can advertise given the Petersburg fork cap (time of writing).
+
+```go
```

Comment:
We should find a format for the test cases that isn't Go.

Reply:
Meh. The Go code is table driven, so it's fairly trivial to copy paste. Whether it's a Go "table" or a markdown table, it makes no difference. You still need to copy paste each cell individually.

From my perspective as an author however, I don't have to screw around with converting Go tables to markdown tables, and then when something changes (like XOR to CRC), I don't have to throw all that work out and redo everything all over.

I.e. It saves time for me, it's more robust against change, and it doesn't matter to anyone else.